torch/csrc/jit/passes/shape_analysis.cpp

```diff
@@ -117,7 +135,28 @@ void PropagateShapeOnNode(Node * node) {
       node->output()->inferTypeFrom(node->t(kvalue));
     } break;
     case kUndefined: {
-      node->output()->setType(nullptr);
+      node->output()->setType(DynamicType::get());
+    } break;
+    case kIf: {
+      auto then_block = node->blocks()[0];
+      auto else_block = node->blocks()[1];
+      PropagateShapeOnBlock(then_block);
+      PropagateShapeOnBlock(else_block);
+      mergeTypes(then_block->outputs(), else_block->outputs(), node->outputs());
+    } break;
+    case kLoop: {
+      auto body_block = node->blocks()[0];
+      // propagate counter type
+      body_block->inputs()[0]->setType(node->inputs()[0]->type());
+      // propagate loop-carried input types to block inputs
+      auto loop_carried_inputs = node->inputs().slice(2); // skip max, cond
+      auto loop_carried_block = body_block->inputs().slice(1); // skip trip
```

Comment:
Do we have any docs describing the input/output structure in the JIT? I don't think so, but we really should have. I see you even changed a few things as part of this PR, and I'm starting to get lost (since our loop is a weird combination of while/for at the same time, and we only have a flat list of phis)

Reply:
I added a comment in compiler.cpp for Loop in particular to describe its fields. We should centralize these somewhere though in the future.